airbyte-integrations/connectors/source-google-search-console/source_google_search_console/config_migrations.py

```diff
@@ -0,0 +1,109 @@
+        >   JSON STR: "{name: my_report}, {dimensions: [a,b,c]}"
+    """
+
+    depercated_message: str = """
+        The `Custom Reports` declared using the deprecated structure of JSON String and are transfomed automatically.
+        Please update the `Custom Reports` structue for this source.
+        Visit: `Sources > select your `Google Search Console` source > Settings > Custom Reports` and update the section using built-in report builder."
+    """
+
+    @classmethod
+    def should_migrate(cls, config: Mapping[str, Any]) -> bool:
+        """
+        This method determines wether or not the config should be migrated to have the new structure for the `custom_reports`
+        Returns:
+            > True, if the transformation is neccessary
+        Or
+            > False, otherwise.
+        Or
+            > Raises the Exception if the structure could not be migrated.
+        """
+
+        if "custom_reports" in config:
+            custom_reports = config["custom_reports"]
+            # check the old structure first
+            if isinstance(custom_reports, str):
+                logger.warning(cls.depercated_message)
+                return True
+            # if the structure is new - bypass the transformation
+            elif isinstance(custom_reports, list):
+                return False
+            # raise an error otherwise
+            else:
+                raise Exception(
+                    f"Custom Reports have invalid structure of {type(custom_reports)}, the structure should be either JSON String or List of Objects."
+                )
+        return False
+
+    @classmethod
+    def transform(cls, source: Source, config: Mapping[str, Any]) -> Mapping[str, Any]:
+        # transfom `json_str` to `list` of objects
+        return source._validate_custom_reports(config)
+
+    @classmethod
+    def migrate(cls, source: Source) -> List[str]:
+        """
+        This method checks the input args, should the config be migrated or not, transform if neccessary.
+
+        Returns:
+            > The list of updated input args, if the transformation was applied to the config
+        Or
+            > The list of original input args, otherwise.
+        """
+        # get input args
+        args = sys.argv[1:]
+        # get config path
+        config_path = AirbyteEntrypoint(source).extract_config(args)
+        # proceed only if `--config` arg is provided
+        if config_path:
+            # read the existing config
+            config = source.read_config(config_path)
+            # migration check
+            if cls.should_migrate(config):
+                # make new path for migrated config
+                migated_config_path = config_path.replace("config", "migrated_config")
+                # modify the config
+                migrated_config = cls.transform(source, config)
+                # save the config
+                source.write_config(migrated_config, migated_config_path)
```

Comment:
we should also emit a CONTROL message to update the config, see the one time use oauth authenticator for an example

Reply:
It's done. The UI gets updated after the CONTROL message comes out.